source/base/mpi.cc

```diff
@@ -812,8 +814,468 @@ namespace Utilities
+            }
+
+          // send dofs to each process
+          for (auto rank_pair : relevant_procs_map)
+            {
+              int rank  = rank_pair.first;
+              int index = rank_pair.second;
+
+              // create index set and compress data to be sent
+              auto &   indices_i = temp[index];
+              IndexSet is(dict.size);
+              is.add_indices(indices_i.begin(), indices_i.end());
+              is.compress();
+
+              // translate index set to a list of pairs
+              auto &buffer = buffers[index];
+              for (auto interval = is.begin_intervals();
+                   interval != is.end_intervals();
+                   interval++)
+                buffer.emplace_back(*interval->begin(), interval->last() + 1);
+
+              // send data
+              MPI_Isend(&buffer[0],
+                        buffer.size() * 2,
+                        DEAL_II_DOF_INDEX_MPI_TYPE,
+                        rank,
+                        StateMachine::tag_setup,
+                        comm,
+                        &request[index]);
+            }
+        }
+
+
+        // 4) receive messages until all dofs in dict are processed
+        while (dict.local_size != dic_local_rececived)
+          {
+            // wait for an incoming massage
+            MPI_Status status;
+            MPI_Probe(MPI_ANY_SOURCE, StateMachine::tag_setup, comm, &status);
+
+            // retrieve size of incoming massage
+            int number_amount;
+            MPI_Get_count(&status, DEAL_II_DOF_INDEX_MPI_TYPE, &number_amount);
+
+            auto other_rank = status.MPI_SOURCE;
+
+            // receive massage
+            std::vector<
+              std::pair<types::global_dof_index, types::global_dof_index>>
+              buffer(number_amount / 2);
+            MPI_Recv(&buffer[0],
+                     number_amount,
+                     DEAL_II_DOF_INDEX_MPI_TYPE,
+                     other_rank,
+                     StateMachine::tag_setup,
+                     comm,
+                     &status);
+
+            // process message: loop over all intervals
+            for (auto interval : buffer)
+              for (types::global_dof_index i = interval.first;
+                   i < interval.second;
+                   i++)
+                {
+                  dict.array[i - dict.local_range.first] = other_rank;
+                  dic_local_rececived++;
+                }
+          }
+
+        // 5) make sure that all messages have been sent
+        std::vector<MPI_Status> status(relevant_procs_size);
+        MPI_Waitall(relevant_procs_size, &request[0], &status[0]);
+      }
+
+
+    } // namespace ComputeIndexOwner
+
+#endif
+
+    std::vector<unsigned int>
+    compute_index_owner(const IndexSet &owned_dofs,
+                        const IndexSet &ghosted_dofs,
+                        const MPI_Comm &comm)
+    {
+#ifdef DEAL_II_WITH_MPI
```

Comment:
Assert that both IndexSets have the same size. What happens in your computation if ghosted_dofs also happen to contain the owned_dofs?

Reply:
Added assert in b634f20b8c95545f0a54928b893d41f826558bef. @tjhei: the explanation to your answer is now in the Doxygen documentation: if the user is interested in the ranks only of the ghosted dofs, he or she should explicitly exclude the locally indices from `ghosted_dofs`.